Add locking support to wolfSSL

cotequeiroz opened the pull request:
This takes advantage of wolfSSL openssl compatibility layer, so all that that's needed are library detection, and inclusion of specific headers.
WolfSSL must be built with --enable-opensslextra to enable the required API, and that's being checked at build time, with a warning if disabled.

This was tested with a WRT3200ACM router (arm) running openwrt master branch, linux 4.14.105, wolfssl 3.15.7, libcurl 4.7.65.1, under a light load.

Signed-off-by: Eneas U de Queiroz <cotequeiroz@gmail.com>


Edit:  python versions tested: 2.7.16, 3.7.2

p commented:
@cotequeiroz Can you rebase this PR on master please?

p commented:
Also would you mind adding wolfssl to documentation wherever other ssl backends are mentioned?

cotequeiroz commented:
> 
> 
> Also would you mind adding wolfssl to documentation wherever other ssl backends are mentioned?

That was an oversight on my part.  I've corrected that and rebased on current master.

p commented:
Thank you!
